mesonbuild/envconfig.py

```diff
@@ -469,11 +471,27 @@ def from_literal(cls, literal: T.Dict[str, str]) -> 'MachineInfo':
     def detect(cls, compilers: T.Optional[CompilersDict] = None) -> MachineInfo:
         return cls(
             detect_system(),
-            detect_cpu_family(compilers) if compilers is not None else None,
-            detect_cpu(compilers) if compilers is not None else None,
+            detect_cpu_family(compilers or {}),
+            detect_cpu(compilers or {}),
             sys.byteorder,
+            compilers is None,
         )
 
+    def redetect(self, compilers: CompilersDict) -> None:
+        """Redetect cpu_family and cpu with current compiler data.
+
+        Sometimes when we've initially detected the cpu_family and cpu we don't
+        have access to the compilers, which we need for getting accurate results.
+        In order to fix that, once a compiler is added we redetect the cpu and
+        cpu_family.
+
+        :param compilers: A dictionary of {lang: compiler}
+        """
+        if self._allow_redetect:
```

Comment:
after we've re-detected once already. Basically we do an initial "guess" even if you don't have any compilers yet, then we go back later once there's a compiler, and try to get more accurate information. We end up running that for both the host and the build in a cross setup, but I'm not convinced that we should be running that for the host compiler ever, since that's ignoring what a user put in their cross file...